cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Lookup.java

```diff
@@ -16,15 +16,43 @@
 
 package co.cask.cdap.etl.api;
 
+import co.cask.cdap.api.data.schema.Schema;
+
+import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
+import javax.annotation.Nullable;
 
 /**
  * Exposes read-only lookup operations on datasets.
  *
- * @param <T> the type of object that will be returned for a lookup
+ * @param <T> the type of object that will be returned for a lookup with String as the key
+ * @param <R> the raw type of object that will be returned for lookups with byte[] as the key
```

Comment:
The problem is that for TableLookup, the key can be byte[] or String, and the value can be Row or StructuredRecord
The KeyValueTableLookup key can be byte[] or String, and the value can be byte[] or String.

Supporting this is difficult with the current Lookup interface. I'm going to update the Lookup interface to not have all of those methods.